presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergAbstractMetadata.java

```diff
@@ -916,39 +903,29 @@ public OptionalLong metadataDelete(ConnectorSession session, ConnectorTableHandl
             throw new TableNotFoundException(handle.getSchemaTableName());
         }
 
-        TableScan scan = icebergTable.newScan();
         TupleDomain<IcebergColumnHandle> domainPredicate = layoutHandle.getValidPredicate();
-
-        if (!domainPredicate.isAll()) {
-            Expression filterExpression = toIcebergExpression(domainPredicate);
-            scan = scan.filter(filterExpression);
-        }
-
-        try (CloseableIterable<FileScanTask> files = scan.planFiles()) {
-            return OptionalLong.of(removeScanFiles(icebergTable, files));
-        }
-        catch (IOException e) {
-            throw new PrestoException(GENERIC_INTERNAL_ERROR, "failed to scan files for delete", e);
-        }
+        return removeScanFiles(icebergTable, domainPredicate);
     }
 
     /**
-     * Deletes all the files within a particular scan
+     * Deletes all the files for a specific predicate
      *
      * @return the number of rows deleted from all files
      */
-    private long removeScanFiles(Table icebergTable, Iterable<FileScanTask> scan)
+    private OptionalLong removeScanFiles(Table icebergTable, TupleDomain<IcebergColumnHandle> predicate)
     {
         transaction = icebergTable.newTransaction();
-        DeleteFiles deletes = transaction.newDelete();
-        AtomicLong rowsDeleted = new AtomicLong(0L);
-        scan.forEach(t -> {
-            deletes.deleteFile(t.file());
-            rowsDeleted.addAndGet(t.estimatedRowsCount());
-        });
-        deletes.commit();
+        DeleteFiles deleteFiles = transaction.newDelete()
+                .deleteFromRowFilter(toIcebergExpression(predicate));
+        deleteFiles.commit();
         transaction.commitTransaction();
-        return rowsDeleted.get();
+
+        Map<String, String> summary = icebergTable.currentSnapshot().summary();
```

Comment:
In theory, it is like this. But considering the optimistic implementation and conflict resolution mechanism of Iceberg's transaction, this situation seems unlikely to occur in reality. Because committing a new transaction between these two closely consecutive actions requires too much effort. For example, at least a manifest list file needs to be rewritten.

At the moment, I think introducing a lock to handle this situation seems a bit too heavy-handed. And as you say, Iceberg seems didn't provide a API to get the current snapshot theoretically competition-safe. So maybe we could just leave it here and consider it as ok. And come back again when the situation really happens. What's your opinoin?

Reply:
I know the likelihood is small, but we don't have control over concurrent processes which may be writing to the table.. I regularly see cases in the iceberg community where folks have pipelines which write to their iceberg tables quite frequently say, once every few seconds. It doesn't seem entirely impossible that we could have something surface where an external pipeline is changing the table in between?

Anyways, I think since the API doesn't provide a proper way to do this, the solution we have now is fine. I just wanted to bring it up.